Add 340 release notes (#4615)
prestosql/presto: presto-docs/src/main/sphinx/release/release-340.md

# Release 339 (XX Jul 2020)

## General Changes

* Add support for query parameters in `LIMIT`, `OFFSET` and `FETCH FIRST` clauses. ({issue}`4529`, {issue}`4601`)

> * Improve dynamic partition pruning for short queries. ({issue}`4388`}

Reviewer: Add 
```
General
* Expose information about query type in query Event Listener. ({issue}`4592`)
```

## Hive Connector Changes

> * Improve performance when querying tables with `skip.header.line.count` property set to 1. ({issue}`4513`)

Reviewer: Should we mention that this only affects textfiles that spans across several splits?

Reviewer 2: like
```
... when querying a table with large files and with `skip.header.line.count` property set to 1.
```

?

Reviewer: 👍

# Release 340 (XX Aug 2020)

## General Changes

* Add support for query parameters in `LIMIT`, `OFFSET` and `FETCH FIRST` clauses. ({issue}`4529`, {issue}`4601`)
* Improve dynamic partition pruning for short queries. ({issue}`4388`}

> * Reduce CPU load by cancelling query stages from which data is no longer required. ({issue}`4290`)

Reviewer 3: I would change it to:
```
Reduce cluster load by ...
```

> # Release 340 (XX Aug 2020)

Reviewer: Please add:
```
Security
* Verify access to referenced table in `CREATE TABLE LIKE`. ({issue}`4472)
```

# Release 340 (XX Aug 2020)

> ## General Changes

Reviewer: Please add:
```
 * Add `IF EXISTS `and `IF NOT EXISTS` syntax to `ALTER TABLE`. ({issue}`4651`)
```

* Add support for query parameters in `LIMIT`, `OFFSET` and `FETCH FIRST` clauses. ({issue}`4529`, {issue}`4601`)
* Add {func}`bitwise_left_shift`, {func}`bitwise_right_shift` and {func}`bitwise_right_shift_arithmetic`. ({issue}`772`)

> * Improve dynamic partition pruning for short queries. ({issue}`4388`}

Reviewer: This is quite vague. What was improved? What kind of improvements user might expect? Shorter latency, lower memory footprint? What does it mean short query?

FYI @sopel39

* Reduce cluster load by cancelling query stages from which data is no longer required. ({issue}`4290`)
* Reduce query memory usage by improving retained size estimation for `VARCHAR` and `CHAR` types. ({issue}`4123`)
* Improve query performance for queries containing ``starts_with`` predicates. ({issue}`4669`)
* Fix failure when `GROUP BY` clause contains duplicate expressions. ({issue}`4609`)
* Fix potential hang during query planning ({issue}`4635`).

## Security

* Verify access to referenced table in `CREATE TABLE LIKE`. ({issue}`4472`)

## JDBC Driver Changes

* Fix handling of dates and timestamps before the 20th century. ({issue}`4563`)

## Elasticsearch Connector Changes

* Fix failure when index mapping is missing. ({issue}`4535`)

> ## Hive Connector Changes

Reviewer: Please add:
```
* Expose hit rate statistics for Hive metastore cache via JMX. ({issue}`4458`)
```

* Improve performance when querying tables with `skip.header.line.count` property set to 1. ({issue}`4513`)
* Allow creating a table with `external_location` when schema's location is not valid. ({issue}`4069`)
* Add read support for tables that were created as non-transactional and converted to be
  transactional later. ({issue}`2293`)

> * Allow creation of transactional tables. Note: inserts to transactional tables

Reviewer: ```suggestion
* Allow `CREATE TABLE` for transactional tables. Note: inserts to transactional tables
```

Reviewer 4: The original is fine. It's explaining a concept, not specific SQL syntax.

# Release 340 (XX Aug 2020)

## General Changes

* Add support for query parameters in `LIMIT`, `OFFSET` and `FETCH FIRST` clauses. ({issue}`4529`, {issue}`4601`)
* Add {func}`bitwise_left_shift`, {func}`bitwise_right_shift` and {func}`bitwise_right_shift_arithmetic`. ({issue}`772`)
* Add `IF EXISTS `and `IF NOT EXISTS` syntax to `ALTER TABLE`. ({issue}`4651`)
* Improve dynamic partition pruning for short queries. ({issue}`4388`}

> * Reduce cluster load by cancelling query stages from which data is no longer required. ({issue}`4290`)

Reviewer 2: "Improve cluster throughput"?

* Reduce query memory usage by improving retained size estimation for `VARCHAR` and `CHAR` types. ({issue}`4123`)

> * Improve query performance for queries containing ``starts_with`` predicates. ({issue}`4669`)

Reviewer 4: This should be
```
* Improve query performance for queries containing {func}`starts_with`. ({issue}`4669`)

* Improve performance of queries that use `DECIMAL` data type. ({issue}`4730`)
* Fix failure when `GROUP BY` clause contains duplicate expressions. ({issue}`4609`)
* Fix potential hang during query planning ({issue}`4635`).

> ## Security

Reviewer 4: Security Changes

> * Verify access to referenced table in `CREATE TABLE LIKE`. ({issue}`4472`)

Reviewer 2: Shouldn;'t this say what is being fixed? like "Fix unprivileged access to table's schema via CREATE TABLE LIKE"?

## JDBC Driver Changes

> * Fix handling of dates and timestamps before the 20th century. ({issue}`4563`)

Reviewer 2: or before "1900-01-01" ?

or maybe even "Fix handling of dates before 1582-10-15" , "Fix handling of timestamps before 1900-01-01" (two bullets)?

## Elasticsearch Connector Changes

* Fix failure when index mapping is missing. ({issue}`4535`)

## Hive Connector Changes

* Improve performance when querying tables with `skip.header.line.count` property set to 1. ({issue}`4513`)
* Allow creating a table with `external_location` when schema's location is not valid. ({issue}`4069`)
* Add read support for tables that were created as non-transactional and converted to be
  transactional later. ({issue}`2293`)

> * Allow creation of transactional tables. Note: inserts to transactional tables

Reviewer 4: Let's make the second sentence

> Note that writing to transactional tables is not yet supported.

  are still not supported. ({issue}`4516`)

> * Add feature toggle (`hive.metastore.glue.max-error-retries`) for configuring number of retries in Glue client. ({issue}`4611`)

Reviewer 4: ```
* Add `hive.metastore.glue.max-error-retries` configuration property for the
  number of retries performed when accessing the Glue metastore. ({issue}`4611`)

> * Add support for Java Key Store (extending existing PEM support) to Thrift Hive meta store TLS configuration. ({issue}`4432`)

Reviewer 4: ```
* Support using Java KeyStore files for Thrift metastore TLS configuration. ({issue}`4432`)

* Expose hit rate statistics for Hive metastore cache via JMX. ({issue}`4458`)

> * Fix query failure when S3 data location contains `_$folder$` Hadoop s3n marker object. ({issue}`4552`)

Reviewer 4: ```
* Fix query failure when S3 data location contains a `_$folder$` marker object. ({issue}`4552`)
```

> * Fix type coercion failure when the schema evolved without type changes. ({issue}`3967`)

Reviewer 2: i don't understand this. eg how did it evolve?

Reviewer 4: I had the same question

Author: Looking at the PR, this happens when the type of the column changed (e.g., from `row(b bigint, c varchar)` to `row(b bigint, c varchar, f int)`, but the type of the field being reference didn't: `b` is still `bigint`.

Author: How about "Fix failure when referencing nested fields of a ROW type when table and partition metadata differs"? It's a bit more accurate, without going into low-level details.

Reviewer: Sounds good to me.

## Kafka Connector Changes

* Add insert support for Raw data format. ({issue}`4417`)
* Add insert support for JSON. ({issue}`4477`)
* Remove unused ``kafka.connect-timeout`` configuration properties. ({issue}`4664`)

## MongoDB Connector Changes

* Add `mongodb.max-connection-idle-time` properties to limit the maximum idle time of a pooled connection. ({issue}`4483`)

## Phoenix Connector Changes

> * Add table level property to specify data block encoding when creating Phoenix tables. ({issue}`4617`)

Reviewer 4: Remove redundant "Phoenix" word

> * Fix `NullPointerException` when listing schemas. ({issue}`4560`)

Reviewer 4: maybe say "Fix query failure"?

## PostgreSQL Connector Changes

> * Push down `count()` aggregation over constant expression (e.g. `count(1)`). ({issue}`4362`)

Reviewer 4: ```
* Push down {func}`count` aggregations over constant expressions.
  For example, `SELECT count(1)`. ({issue}`4362`)

# Release 340 (XX Aug 2020)

## General Changes

* Add support for query parameters in `LIMIT`, `OFFSET` and `FETCH FIRST` clauses. ({issue}`4529`, {issue}`4601`)
* Add experimental support for recursive queries. ({issue}`4250`)
* Add {func}`bitwise_left_shift`, {func}`bitwise_right_shift` and {func}`bitwise_right_shift_arithmetic`. ({issue}`772`)
* Add {func}`luhn_check`. ({issue}`4011`)
* Add `IF EXISTS `and `IF NOT EXISTS` syntax to `ALTER TABLE`. ({issue}`4651`)
* Include remote host in error info for page transport errors. ({issue}`4511`)

> * Enable HTTP/2 by default for internal communication. This can be disabled using the

Reviewer 4: Remove this since we're disabling HTTP/2

  `internal-communication.http2.enabled` configuration property. ({issue}`4696`)
* Improve minimum latency for dynamic partition pruning. ({issue}`4388`}
* Reduce cluster load by cancelling query stages from which data is no longer required. ({issue}`4290`)
* Reduce query memory usage by improving retained size estimation for `VARCHAR` and `CHAR` types. ({issue}`4123`)
* Improve query performance for queries containing {func}`starts_with`. ({issue}`4669`)
* Improve performance of queries that use `DECIMAL` data type. ({issue}`4730`)
* Fix failure when `GROUP BY` clause contains duplicate expressions. ({issue}`4609`)
* Fix potential hang during query planning ({issue}`4635`).

## Security Changes

* Fix unprivileged access to table's schema via `CREATE TABLE LIKE`. ({issue}`4472`)

## JDBC Driver Changes

> * Fix handling of dates and timestamps before the 20th century. ({issue}`4563`)

Reviewer 5: Is this redundant with the two items below?

Author: Yes, forgot to remove the old one.

* Allow creating a table with `external_location` when schema's location is not valid. ({issue}`4069`)
* Add read support for tables that were created as non-transactional and converted to be
  transactional later. ({issue}`2293`)
* Allow creation of transactional tables. Note that writing to transactional tables 
  is not yet supported. ({issue}`4516`)
* Add `hive.metastore.glue.max-error-retries` configuration property for the
  number of retries performed when accessing the Glue metastore. ({issue}`4611`)
* Support using Java KeyStore files for Thrift metastore TLS configuration. ({issue}`4432`)
* Expose hit rate statistics for Hive metastore cache via JMX. ({issue}`4458`)
* Improve performance when querying a table with large files and with `skip.header.line.count` property set to 1. ({issue}`4513`)
* Improve performance of reading JSON tables. ({issue}`4705`)
* Fix query failure when S3 data location contains a `_$folder$` marker object. ({issue}`4552`)
* Fix type coercion failure when the schema evolved without type changes. ({issue}`3967`)

## Kafka Connector Changes

* Add insert support for Raw data format. ({issue}`4417`)
* Add insert support for JSON. ({issue}`4477`)

> * Remove unused ``kafka.connect-timeout`` configuration properties. ({issue}`4664`)

Reviewer 4: Single backtick
